// @cloudbase/oauth/packages/oauth/dist/cjs/auth/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorType = exports.VerificationUsages = exports.ApiUrlsV2 = exports.ApiUrls = void 0;
var ApiUrls;
(function (ApiUrls) {
    ApiUrls["AUTH_SIGN_IN_URL"] = "/auth/v1/signin";
    ApiUrls["AUTH_SIGN_IN_ANONYMOUSLY_URL"] = "/auth/v1/signin/anonymously";
    ApiUrls["AUTH_SIGN_IN_WITH_PROVIDER_URL"] = "/auth/v1/signin/with/provider";
    ApiUrls["AUTH_SIGN_IN_WITH_WECHAT_URL"] = "/auth/v1/signin/wechat/noauth";
    ApiUrls["AUTH_SIGN_UP_URL"] = "/auth/v1/signup";
    ApiUrls["AUTH_TOKEN_URL"] = "/auth/v1/token";
    ApiUrls["AUTH_REVOKE_URL"] = "/auth/v1/revoke";
    ApiUrls["PROVIDER_BIND_URL"] = "/auth/v1/user/provider/bind";
    ApiUrls["PROVIDER_TOKEN_URL"] = "/auth/v1/provider/token";
    ApiUrls["PROVIDER_URI_URL"] = "/auth/v1/provider/uri";
    ApiUrls["USER_ME_URL"] = "/auth/v1/user/me";
    ApiUrls["WEDA_USER_URL"] = "/auth/v1/plugin/weda/userinfo";
    ApiUrls["USER_QUERY_URL"] = "/auth/v1/user/query";
    ApiUrls["USER_PRIFILE_URL"] = "/auth/v1/user/profile";
    ApiUrls["USER_BASIC_EDIT_URL"] = "/auth/v1/user/basic/edit";
    ApiUrls["USER_TRANS_BY_PROVIDER_URL"] = "/auth/v1/user/trans/by/provider";
    ApiUrls["VERIFICATION_URL"] = "/auth/v1/verification";
    ApiUrls["VERIFY_URL"] = "/auth/v1/verification/verify";
    ApiUrls["CAPTCHA_DATA_URL"] = "/auth/v1/captcha/data";
    ApiUrls["VERIFY_CAPTCHA_DATA_URL"] = "/auth/v1/captcha/data/verify";
    ApiUrls["GET_CAPTCHA_URL"] = "/auth/v1/captcha/init";
    ApiUrls["PROVIDER_LIST"] = "/auth/v1/user/provider";
    ApiUrls["PROVIDER_UNBIND_URL"] = "/auth/v1/user/provider";
    ApiUrls["CHECK_PWD_URL"] = "/auth/v1/user/sudo";
    ApiUrls["SUDO_URL"] = "/auth/v1/user/sudo";
    ApiUrls["BIND_CONTACT_URL"] = "/auth/v1/user/contact";
    ApiUrls["AUTH_SET_PASSWORD"] = "/auth/v1/user/password";
    ApiUrls["AUTH_RESET_PASSWORD"] = "/auth/v1/reset";
    ApiUrls["AUTH_GET_DEVICE_CODE"] = "/auth/v1/device/code";
    ApiUrls["CHECK_USERNAME"] = "/auth/v1/checkUsername";
    ApiUrls["CHECK_IF_USER_EXIST"] = "/auth/v1/checkIfUserExist";
    ApiUrls["GET_PROVIDER_TYPE"] = "/auth/v1/mgr/provider/providerSubType";
    ApiUrls["GET_MINIPROGRAM_QRCODE"] = "/auth/v1/qrcode/generate";
    ApiUrls["GET_MINIPROGRAM_QRCODE_STATUS"] = "/auth/v1/qrcode/get/status";
})(ApiUrls = exports.ApiUrls || (exports.ApiUrls = {}));
var ApiUrlsV2;
(function (ApiUrlsV2) {
    ApiUrlsV2["AUTH_SIGN_IN_URL"] = "/auth/v2/signin/username";
    ApiUrlsV2["AUTH_TOKEN_URL"] = "/auth/v2/token";
    ApiUrlsV2["USER_ME_URL"] = "/auth/v2/user/me";
    ApiUrlsV2["VERIFY_URL"] = "/auth/v2/signin/verificationcode";
    ApiUrlsV2["AUTH_SIGN_IN_WITH_PROVIDER_URL"] = "/auth/v2/signin/with/provider";
    ApiUrlsV2["AUTH_PUBLIC_KEY"] = "/auth/v2/signin/publichkey";
})(ApiUrlsV2 = exports.ApiUrlsV2 || (exports.ApiUrlsV2 = {}));
var VerificationUsages;
(function (VerificationUsages) {
    VerificationUsages["REGISTER"] = "REGISTER";
    VerificationUsages["SIGN_IN"] = "SIGN_IN";
    VerificationUsages["PASSWORD_RESET"] = "PASSWORD_RESET";
    VerificationUsages["EMAIL_ADDRESS_CHANGE"] = "EMAIL_ADDRESS_CHANGE";
    VerificationUsages["PHONE_NUMBER_CHANGE"] = "PHONE_NUMBER_CHANGE";
})(VerificationUsages = exports.VerificationUsages || (exports.VerificationUsages = {}));
var ErrorType;
(function (ErrorType) {
    ErrorType["INVALID_ARGUMENT"] = "invalid_argument";
    ErrorType["DEADLINE_EXCEEDED"] = "deadline_exceeded";
    ErrorType["NOT_FOUND"] = "not_found";
    ErrorType["ALREADY_EXISTS"] = "already_exists";
    ErrorType["PERMISSION_DENIED"] = "permission_denied";
    ErrorType["ABORTED"] = "aborted";
    ErrorType["OUT_OF_RANGE"] = "out_of_range";
    ErrorType["UNIMPLEMENTED"] = "unimplemented";
    ErrorType["INTERNAL"] = "internal";
    ErrorType["UNAVAILABLE"] = "unavailable";
    ErrorType["DATA_LOSS"] = "data_loss";
    ErrorType["CAPTCHA_REQUIRED"] = "captcha_required";
    ErrorType["CAPTCHA_INVALID"] = "captcha_invalid";
    ErrorType["INVALID_PASSWORD"] = "invalid_password";
    ErrorType["PASSWORD_NOT_SET"] = "password_not_set";
    ErrorType["INVALID_STATUS"] = "invalid_status";
    ErrorType["USER_PENDING"] = "user_pending";
    ErrorType["USER_BLOCKED"] = "user_blocked";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
