// @cloudbase/oauth/dist/auth/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceAuthorizeState = exports.ErrorType = exports.VerificationUsages = exports.ApiUrls = void 0;
var ApiUrls;
(function (ApiUrls) {
    ApiUrls["AUTH_SIGN_IN_URL"] = "/auth/v1/signin";
    ApiUrls["AUTH_SIGN_IN_ANONYMOUSLY_URL"] = "/auth/v1/signin/anonymously";
    ApiUrls["AUTH_SIGN_IN_WITH_PROVIDER_URL"] = "/auth/v1/signin/with/provider";
    ApiUrls["AUTH_SIGN_IN_WITH_WECHAT_URL"] = "/auth/v1/signin/wechat/noauth";
    ApiUrls["AUTH_SIGN_UP_URL"] = "/auth/v1/signup";
    ApiUrls["AUTH_TOKEN_URL"] = "/auth/v1/token";
    ApiUrls["AUTH_REVOKE_URL"] = "/auth/v1/revoke";
    ApiUrls["AUTH_SIGNOUT_URL"] = "/auth/v1/user/signout";
    ApiUrls["PROVIDER_BIND_URL"] = "/auth/v1/user/provider/bind";
    ApiUrls["PROVIDER_TOKEN_URL"] = "/auth/v1/provider/token";
    ApiUrls["PROVIDER_URI_URL"] = "/auth/v1/provider/uri";
    ApiUrls["USER_ME_URL"] = "/auth/v1/user/me";
    ApiUrls["RESET_PASSWORD_URL"] = "/auth/v1/reset";
    ApiUrls["USER_QUERY_URL"] = "/auth/v1/user/query";
    ApiUrls["USER_PROFILE_URL"] = "/auth/v1/user/profile";
    ApiUrls["USER_BASIC_EDIT_URL"] = "/auth/v1/user/basic/edit";
    ApiUrls["USER_TRANS_BY_PROVIDER_URL"] = "/auth/v1/user/trans/by/provider";
    ApiUrls["VERIFICATION_URL"] = "/auth/v1/verification";
    ApiUrls["VERIFY_URL"] = "/auth/v1/verification/verify";
    ApiUrls["CAPTCHA_DATA_URL"] = "/auth/v1/captcha/data";
    ApiUrls["VERIFY_CAPTCHA_DATA_URL"] = "/auth/v1/captcha/data/verify";
    ApiUrls["GET_CAPTCHA_URL"] = "/auth/v1/captcha/init";
    ApiUrls["PROVIDER_LIST_URL"] = "/auth/v1/user/provider";
    ApiUrls["PROVIDER_UNBIND_URL"] = "/auth/v1/user/provider";
    ApiUrls["SUDO_URL"] = "/auth/v1/user/sudo";
    ApiUrls["EDIT_CONTACT_URL"] = "/auth/v1/user/contact";
    ApiUrls["AUTH_SET_PASSWORD_URL"] = "/auth/v1/user/password";
    ApiUrls["CHECK_IF_USER_EXIST_URL"] = "/auth/v1/checkIfUserExist";
    ApiUrls["AUTHORIZE_URL"] = "/auth/v1/user/authorize";
    ApiUrls["AUTHORIZE_DEVICE_URL"] = "/auth/v1/user/device/authorize";
    ApiUrls["AUTHORIZE_DEVICE_CODE_URL"] = "/auth/v1/device/code";
    ApiUrls["AUTHORIZE_INFO_URL"] = "/auth/v1/user/authorize/info";
    ApiUrls["AUTHORIZED_URL"] = "/auth/v1/user/authorized";
    ApiUrls["AUTHORIZED_DELETE_URL"] = "/auth/v1/user/authorized/";
    ApiUrls["AUTHORIZED_DEVICES_URL"] = "/auth/v1/user/authorized/devices";
    ApiUrls["AUTHORIZED_DEVICES_DELETE_URL"] = "/auth/v1/user/authorized/devices/";
    ApiUrls["AUTH_REVOKE_ALL_URL"] = "/auth/v1/user/revoke/all";
    ApiUrls["GET_PROVIDER_TYPE"] = "/auth/v1/mgr/provider/providerSubType";
    ApiUrls["GET_MINIPROGRAM_QRCODE"] = "/auth/v1/qrcode/generate";
    ApiUrls["GET_MINIPROGRAM_QRCODE_STATUS"] = "/auth/v1/qrcode/get/status";
})(ApiUrls = exports.ApiUrls || (exports.ApiUrls = {}));
var VerificationUsages;
(function (VerificationUsages) {
    VerificationUsages["REGISTER"] = "REGISTER";
    VerificationUsages["SIGN_IN"] = "SIGN_IN";
    VerificationUsages["PASSWORD_RESET"] = "PASSWORD_RESET";
    VerificationUsages["EMAIL_ADDRESS_CHANGE"] = "EMAIL_ADDRESS_CHANGE";
    VerificationUsages["PHONE_NUMBER_CHANGE"] = "PHONE_NUMBER_CHANGE";
})(VerificationUsages = exports.VerificationUsages || (exports.VerificationUsages = {}));
var ErrorType;
(function (ErrorType) {
    ErrorType["INVALID_ARGUMENT"] = "invalid_argument";
    ErrorType["DEADLINE_EXCEEDED"] = "deadline_exceeded";
    ErrorType["NOT_FOUND"] = "not_found";
    ErrorType["ALREADY_EXISTS"] = "already_exists";
    ErrorType["PERMISSION_DENIED"] = "permission_denied";
    ErrorType["ABORTED"] = "aborted";
    ErrorType["OUT_OF_RANGE"] = "out_of_range";
    ErrorType["UNIMPLEMENTED"] = "unimplemented";
    ErrorType["INTERNAL"] = "internal";
    ErrorType["UNAVAILABLE"] = "unavailable";
    ErrorType["DATA_LOSS"] = "data_loss";
    // CommonError
    ErrorType["CAPTCHA_REQUIRED"] = "captcha_required";
    ErrorType["CAPTCHA_INVALID"] = "captcha_invalid";
    ErrorType["INVALID_PASSWORD"] = "invalid_password";
    ErrorType["PASSWORD_NOT_SET"] = "password_not_set";
    ErrorType["INVALID_STATUS"] = "invalid_status";
    ErrorType["USER_PENDING"] = "user_pending";
    ErrorType["USER_BLOCKED"] = "user_blocked";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
var DeviceAuthorizeState;
(function (DeviceAuthorizeState) {
    // 完成
    DeviceAuthorizeState["ACCOMPLISHED"] = "ACCOMPLISHED";
    // 等待连接
    DeviceAuthorizeState["WAITING_CONNECT"] = "WAITING_CONNECT";
    // 等待用户点击同意界面
    DeviceAuthorizeState["WAITING_CONSENT"] = "WAITING_CONSENT";
    // 用户拒绝授权
    DeviceAuthorizeState["ACCESS_DENIED"] = "ACCESS_DENIED";
})(DeviceAuthorizeState = exports.DeviceAuthorizeState || (exports.DeviceAuthorizeState = {}));
