// @cloudbase/weda-client/dist/cjs/mp/auth/uitl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuthConfig = exports.redirectToLogin = exports.cleanAccessPermissionCache = exports.getAccessPermission = exports.findLoginPage = exports.checkAnonymous = void 0;
const request_1 = require("../../common/models/request");
const app_1 = require("../app");
const user_1 = require("./user");
async function checkAnonymous() {
    let isAnonymous = true;
    try {
        const { accessToken } = await (0, user_1.getAccessToken)();
        isAnonymous = !accessToken;
        // 有 token 的情况下需要验证匿名登录需要多验证scope === 'anonymous'
        if (accessToken) {
            const scope = await (0, user_1.loginScope)();
            isAnonymous = scope === 'anonymous';
        }
    }
    catch (e) { }
    return isAnonymous;
}
exports.checkAnonymous = checkAnonymous;
function findLoginPage() {
    const { app } = (0, app_1.getWedaAPI)();
    const { pages = [], loginConfigVersion } = app.__internal__.getConfig();
    const custom = pages.find((item) => item.type === 'login');
    if (loginConfigVersion) {
        return (custom || {
            id: 'login',
            packageName: '$wd_system',
            uuid: '$wd_system/login',
            type: 'login',
        });
    }
}
exports.findLoginPage = findLoginPage;
let _AUTH_CACHE_MAP = {};
async function getAccessPermission(appId, packageName = '', pageId, isAdminPortal = false) {
    let cacheKey = `${appId}-${pageId}`;
    if (packageName) {
        const matched = packageName.match(/packages\/(.*)$/);
        const subKey = matched?.[1] || packageName;
        cacheKey = `${appId}-${subKey}-${pageId}`;
    }
    if (_AUTH_CACHE_MAP[cacheKey] !== undefined) {
        return _AUTH_CACHE_MAP[cacheKey];
    }
    const res = await (0, request_1.callWedaApiFn)('DescribeResourcesPermission', {
        ResourceType: isAdminPortal ? 'modelApp' : 'app',
        ResourceIdList: [cacheKey],
        AppResourceId: appId,
    });
    if (Array.isArray(res)) {
        const resourceInfo = res.find((i) => i.ResourceId === cacheKey);
        const isAccess = !!resourceInfo?.IsAccess;
        _AUTH_CACHE_MAP[cacheKey] = {
            isAccess,
            roleId: resourceInfo?.RoleId,
        };
        return _AUTH_CACHE_MAP[cacheKey];
    }
    else {
        throw new Error(`鉴权接口返回参数不正确 - ${JSON.stringify(res)}`);
    }
}
exports.getAccessPermission = getAccessPermission;
function cleanAccessPermissionCache() {
    _AUTH_CACHE_MAP = {};
}
exports.cleanAccessPermissionCache = cleanAccessPermissionCache;
function redirectToLogin(currentPage = undefined) {
    // 去登录则清空权限缓存。
    cleanAccessPermissionCache();
    const { app } = (0, app_1.getWedaAPI)();
    const loginPage = findLoginPage();
    const home = app?.__internal__?.getConfig()?.pages?.[0];
    if (!currentPage) {
        currentPage = app.utils.getCurrentPage() || { id: home?.id };
    }
    const loginPageUUID = loginPage?.uuid || loginPage?.id;
    const currentUUID = currentPage.uuid || currentPage.id;
    if (loginPageUUID === currentUUID) {
        return true;
    }
    if (loginPage) {
        app.redirectTo({
            pageId: loginPage.id,
            packageName: loginPage.packageName,
            params: {
                sourcePagePackageName: currentPage.__internal__.packageName || undefined,
                sourcePageId: currentPage.id,
                sourcePageParams: currentPage.params || currentPage.dataset?.params,
            },
        });
    }
    else {
        app.showToast({
            title: '用户未登录',
            icon: 'error',
        });
    }
}
exports.redirectToLogin = redirectToLogin;
let _AUTH_CONFIG_CACHE = null;
async function getAuthConfig(isAdminPortal = false) {
    const { app } = (0, app_1.getWedaAPI)();
    if (_AUTH_CONFIG_CACHE) {
        return _AUTH_CONFIG_CACHE;
    }
    const { envId, accessType } = app.__internal__.getConfig();
    try {
        const res = await app.cloud.callWedaApi({
            action: 'DescribeRuntimeResourceStrategy',
            data: {
                ResourceType: isAdminPortal ? 'modelApp' : 'app',
                ResourceId: accessType === 'env' && !isAdminPortal ? envId : app.id,
            },
        });
        const settingData = {};
        // 云api不支持map只能传字符串，需要转换
        res.forEach((item) => {
            settingData[item.Key] = ['AllowRegister', 'NeedLogin'].includes(item.Key) ? item.Value === '1' : item.Value;
        });
        _AUTH_CONFIG_CACHE = settingData;
        return _AUTH_CONFIG_CACHE;
    }
    catch (e) {
        return {
            NeedLogin: false,
            RejectStrategy: 'show_warning',
        };
    }
}
exports.getAuthConfig = getAuthConfig;
