// @cloudbase/oauth/src/oauthclient/models.ts
import {RequestFn, RequestOptions, Storage} from "../app";
import {ErrorType} from './consts';

// Credentials
export interface Credentials {
    token_type?: string;
    access_token?: string;
    refresh_token?: string;
    scope?: string;
    expires_in?: number;
    expires_at?: Date;
    sub?: string;
}

// An Error For all concern
export interface ResponseError {
    error: ErrorType;
    error_description?: string;
    error_uri?: string;
    details?: any;
    request_id?: string;
}

// refer https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Authentication
export interface AuthClientRequestOptions extends RequestOptions {
    headers?: {
        'x-request-id'?: string;
        [key: string]: any;
    };
    // Bearer https://datatracker.ietf.org/doc/html/rfc6750
    withCredentials?: boolean;
    // Basic https://datatracker.ietf.org/doc/html/rfc7617
    withBasicAuth?: boolean;
    retry?: number;

    [key: string]: any;
}

export interface OAuth2ClientOptions {
    clientId: string;  // Storage, default is localStorage, setItem(k, v), getItem(k),removeItem(k)
    storage: Storage;
    request: RequestFn;
    clientSecret?: string;
    // default value is 1,min value is 0, max value is 5
    retry?: number;
    refreshTokenFunc?: (refreshToken?: string) => Promise<Credentials>;
    // set the token in url query instead of header
    tokenInURL?: boolean;
    headers?: { [key: string]: string };
}
