// @cloudbase/oauth/packages/oauth/dist/esm/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { OAuth2Client } from './oauth2client/oauth2client';
import { Auth } from './auth/apis';
import * as authModels from './auth/models';
export { Syntax, ErrorType } from './oauth2client/consts';
export { defaultStorage, defaultRequest, toResponseError, generateRequestId, OAuth2Client, } from './oauth2client/oauth2client';
export { AuthClient } from './oauth2client/interface';
export { Auth } from './auth/apis';
export { authModels };
var CloudbaseOAuth = (function () {
    function CloudbaseOAuth(authOptions) {
        var apiOrigin = authOptions.apiOrigin, clientId = authOptions.clientId, storage = authOptions.storage, request = authOptions.request, baseRequest = authOptions.baseRequest, anonymousSignInFunc = authOptions.anonymousSignInFunc;
        this.oauth2client = new OAuth2Client({
            apiOrigin: apiOrigin,
            clientId: clientId,
            storage: storage,
            baseRequest: baseRequest || request,
            anonymousSignInFunc: anonymousSignInFunc,
        });
        this.authApi = new Auth(__assign(__assign({ credentialsClient: this.oauth2client }, authOptions), { request: request ? this.oauth2client.request.bind(this.oauth2client) : undefined }));
    }
    return CloudbaseOAuth;
}());
export { CloudbaseOAuth };
