// @cloudbase/oauth/packages/container/dist/esm/utils/index.js
import { constants } from '@cloudbase/utilities';
import { COMPONENT_NAME } from '../constants';
var ERRORS = constants.ERRORS;
export function parseURL(_url) {
    if (document) {
        var a = document.createElement('a');
        a.href = _url;
        return a;
    }
    return new URL(new Request(_url).url);
}
export function isBuffer(buf) {
    return buf instanceof ArrayBuffer || ArrayBuffer.isView(buf);
}
function object2Qs(data) {
    var list = Object.entries(data || {}).reduce(function (list, _a) {
        var key = _a[0], value = _a[1];
        list.push("".concat(encodeURIComponent(key), "=").concat(encodeURIComponent(value)));
        return list;
    }, []);
    if (list.length) {
        return list.join('&');
    }
}
export function serializeRequestBody(headers, data, method) {
    if (method === void 0) { method = 'POST'; }
    var res = data;
    var contentTypeKey = 'Content-Type';
    if (typeof data === 'object' && !isBuffer(data)) {
        if (!headers[contentTypeKey] && !headers['content-type']) {
            headers[contentTypeKey] = 'application/json';
        }
        var contentType = headers[contentTypeKey] || headers['content-type'];
        if (method === 'GET') {
            try {
                return object2Qs(data) || '';
            }
            catch (e) {
                throw new Error(JSON.stringify({
                    code: ERRORS.INVALID_PARAMS,
                    msg: "[".concat(COMPONENT_NAME, ".callContainer] invalid data in query method, ").concat(e.message),
                }));
            }
        }
        switch (contentType) {
            case 'application/json': {
                try {
                    res = JSON.stringify(data);
                }
                catch (e) {
                    throw new Error(JSON.stringify({
                        code: ERRORS.INVALID_PARAMS,
                        msg: "[".concat(COMPONENT_NAME, ".callContainer] invalid data with content-type: application/json, ").concat(e.message),
                    }));
                }
                break;
            }
            case 'application/x-www-form-urlencoded': {
                try {
                    var qs_1 = object2Qs(data);
                    if (qs_1) {
                        res = qs_1;
                    }
                }
                catch (e) {
                    throw new Error(JSON.stringify({
                        code: ERRORS.INVALID_PARAMS,
                        msg: "[".concat(COMPONENT_NAME, ".callContainer] invalid data with content-type: application/x-www-form-urlencoded,").concat(e.message),
                    }));
                }
                break;
            }
        }
    }
    return res;
}
