// @cloudbase/oauth/dist/esm/auth/consts.js
export var ApiUrls;
(function (ApiUrls) {
    ApiUrls["AUTH_SIGN_IN_URL"] = "/auth/v1/signin";
    ApiUrls["AUTH_SIGN_IN_ANONYMOUSLY_URL"] = "/auth/v1/signin/anonymously";
    ApiUrls["AUTH_SIGN_IN_WITH_PROVIDER_URL"] = "/auth/v1/signin/with/provider";
    ApiUrls["AUTH_SIGN_UP_URL"] = "/auth/v1/signup";
    ApiUrls["AUTH_TOKEN_URL"] = "/auth/v1/token";
    ApiUrls["AUTH_REVOKE_URL"] = "/auth/v1/revoke";
    ApiUrls["PROVIDER_BIND_URL"] = "/auth/v1/user/provider/bind";
    ApiUrls["PROVIDER_TOKEN_URL"] = "/auth/v1/provider/token";
    ApiUrls["PROVIDER_URI_URL"] = "/auth/v1/provider/uri";
    ApiUrls["USER_ME_URL"] = "/auth/v1/user/me";
    ApiUrls["USER_QUERY_URL"] = "/auth/v1/user/query";
    ApiUrls["USER_PRIFILE_URL"] = "/auth/v1/user/profile";
    ApiUrls["USER_TRANS_BY_PROVIDER_URL"] = "/auth/v1/user/trans/by/provider";
    ApiUrls["VERIFICATION_URL"] = "/auth/v1/verification";
    ApiUrls["VERIFY_URL"] = "/auth/v1/verification/verify";
    ApiUrls["PROVIDER_LIST"] = "/auth/v1/user/provider";
    ApiUrls["PROVIDER_UNBIND_URL"] = "/auth/v1/user/provider";
    ApiUrls["CHECK_PWD_URL"] = "/auth/v1/user/sudo";
    ApiUrls["SUDO_URL"] = "/auth/v1/user/sudo";
    ApiUrls["BIND_CONTACT_URL"] = "/auth/v1/user/contact";
    ApiUrls["AUTH_SET_PASSWORD"] = "/auth/v1/user/password";
    ApiUrls["AUTH_RESET_PASSWORD"] = "/auth/v1/reset";
    ApiUrls["AUTH_GET_DEVICE_CODE"] = "/auth/v1/device/code";
    ApiUrls["CHECK_USERNAME"] = "/auth/v1/checkUsername";
    ApiUrls["CHECK_IF_USER_EXIST"] = "/auth/v1/checkIfUserExist";
})(ApiUrls || (ApiUrls = {}));
export var VerificationUsages;
(function (VerificationUsages) {
    VerificationUsages["REGISTER"] = "REGISTER";
    VerificationUsages["SIGN_IN"] = "SIGN_IN";
    VerificationUsages["PASSWORD_RESET"] = "PASSWORD_RESET";
    VerificationUsages["EMAIL_ADDRESS_CHANGE"] = "EMAIL_ADDRESS_CHANGE";
    VerificationUsages["PHONE_NUMBER_CHANGE"] = "PHONE_NUMBER_CHANGE";
})(VerificationUsages || (VerificationUsages = {}));
export var ErrorType;
(function (ErrorType) {
    ErrorType["INVALID_ARGUMENT"] = "invalid_argument";
    ErrorType["DEADLINE_EXCEEDED"] = "deadline_exceeded";
    ErrorType["NOT_FOUND"] = "not_found";
    ErrorType["ALREADY_EXISTS"] = "already_exists";
    ErrorType["PERMISSION_DENIED"] = "permission_denied";
    ErrorType["ABORTED"] = "aborted";
    ErrorType["OUT_OF_RANGE"] = "out_of_range";
    ErrorType["UNIMPLEMENTED"] = "unimplemented";
    ErrorType["INTERNAL"] = "internal";
    ErrorType["UNAVAILABLE"] = "unavailable";
    ErrorType["DATA_LOSS"] = "data_loss";
    ErrorType["CAPTCHA_REQUIRED"] = "captcha_required";
    ErrorType["CAPTCHA_INVALID"] = "captcha_invalid";
    ErrorType["INVALID_PASSWORD"] = "invalid_password";
    ErrorType["PASSWORD_NOT_SET"] = "password_not_set";
    ErrorType["INVALID_STATUS"] = "invalid_status";
    ErrorType["USER_PENDING"] = "user_pending";
    ErrorType["USER_BLOCKED"] = "user_blocked";
})(ErrorType || (ErrorType = {}));
